Gems/ROS2ScriptIntegration/Code/Source/Clients/ROS2ScriptIntegrationSystemComponent.cpp

```diff
@@ -0,0 +1,481 @@
+            std::lock_guard<AZStd::mutex> lock(m_subscribersMapMutex);
+            m_subscribers[topicName] = subscriber;
+        }
+    };
+
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<MessageType>> ROS2ScriptIntegrationSystemComponent::GetOrCreatePublisher(
+        const AZStd::string& topicName)
+    {
+        std::lock_guard<AZStd::mutex> lock(m_publisherMapMutex);
+        auto it = m_publishers.find(topicName);
+        if (it != m_publishers.end())
+        {
+            return std::dynamic_pointer_cast<rclcpp::Publisher<MessageType>>(it->second);
+        }
+        auto ros2Node = ROS2::ROS2Interface::Get()->GetNode();
+        if (ros2Node)
+        {
+            // create a new publisher
+            std::shared_ptr<rclcpp::Publisher<MessageType>> publisher =
+                ros2Node->create_publisher<MessageType>(std::string(topicName.c_str()), 10);
+            m_publishers[topicName] = publisher;
+            return publisher;
+        }
+        return nullptr;
+    }
+
+    // Explicit template instantiation for all message types (Publishers)
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::String>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Empty>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Bool>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Float32>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::Int32>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<std_msgs::msg::UInt32>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Point32>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Quaternion>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Transform>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Twist>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Vector3>> GetOrCreatePublisher(const AZStd::string& topicName);
+    template<typename MessageType>
+    std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::PoseStamped>> GetOrCreatePublisher(const AZStd::string& topicName);
+
```

Comment:
explicit template instantiation is not needed if all calls to template methods are in the same *.cpp file as template function definition.